server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -201,55 +211,42 @@ private SimpleTopDocsCollectorContext(IndexReader reader,
                                               boolean hasFilterCollector) throws IOException {
             super(REASON_SEARCH_TOP_HITS, numHits);
             this.sortAndFormats = sortAndFormats;
+
+            // implicit total hit counts are valid only when there is no filter collector in the chain
+            final int hitCount = hasFilterCollector ? -1 : shortcutTotalHitCount(reader, query);
+            final TopDocsCollector<?> topDocsCollector;
+            if (hitCount == -1 && trackTotalHits) {
+                topDocsCollector = createCollector(sortAndFormats, numHits, searchAfter, Integer.MAX_VALUE);
+                topDocsSupplier = new CachedSupplier<>(topDocsCollector::topDocs);
+                totalHitsSupplier = () -> topDocsSupplier.get().totalHits;
+            } else {
+                topDocsCollector = createCollector(sortAndFormats, numHits, searchAfter, 1); // don't compute hit counts via the collector
+                topDocsSupplier = new CachedSupplier<>(topDocsCollector::topDocs);
+                if (hitCount == -1) {
+                    assert trackTotalHits == false;
+                    totalHitsSupplier = () -> new TotalHits(topDocsSupplier.get().scoreDocs.length,
```

Comment:
Let's return `new TotalHits(0,  TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO)` for now ? It is translated to `-1` by the coordinating node anyway since `track_total_hits` is `false` in this case. We also need to handle bwc for mixed search with 6x node. Currently we have an [exception](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/Lucene.java#L403) if the hit count is greater than 0 and not an exact count so I suspect that this change could break some bwc tests randomly. We can revise this when we add the support for `track_total_hits` as a number ?